stdlib/public/core/StringGuts.swift

```diff
@@ -244,859 +248,508 @@ extension _StringGuts {
   @_versioned
-  init(_asciiPointer ptr: UnsafeRawPointer, codeUnitCount: Int) {
-    _sanityCheck(codeUnitCount >= 0)
-    self.init(
-      _bridgeObject(taggingPayload: UInt(bitPattern: ptr)),
-      UInt(bitPattern: codeUnitCount))
+  internal
+  func _ephemeralCocoaString() -> _CocoaString {
+    if _fastPath(_isNative) {
+      return _nativeRawStorage
+    } else if _isNonTaggedCocoa {
+      return _nonTaggedCocoaObject
+    } else if _isSmallCocoa {
+      return _taggedCocoaObject
+    } else if isASCII {
+      return _NSContiguousString(
+        _StringGuts(_asUnmanaged(of: UInt8.self)))
+    } else {
+      return _NSContiguousString(
+        _StringGuts(_asUnmanaged(of: UTF16.CodeUnit.self)))
+    }
   }
 
-  //
-  // Tagged Cocoa Strings
-  //
-
-  ///*fileprivate*/ internal // TODO: private in Swift 4
+  /// Return an NSString instance containing a slice of this string.
+  /// The returned object may contain unmanaged pointers into the
+  /// storage of this string; you are responsible for ensuring that
+  /// it will not outlive `self`.
   @_inlineable
-  public // @testable
-  var _isSmallCocoa: Bool {
-    @inline(__always)
-    get { return _isTagged && _objectBitPattern & _smallBit != 0 }
-  }
-
-  /*fileprivate*/ internal // TODO: private in Swift 4
-  var _smallCocoa: SmallCocoaString? {
-    guard _isSmallCocoa else { return nil }
-    return SmallCocoaString(_otherBits)
+  @_versioned
+  internal
+  func _ephemeralCocoaString(_ range: Range<Int>) -> _CocoaString {
+    if _slowPath(_isOpaque) {
+      return _asOpaque()[range].cocoaSlice()
+    } else {
+      return _NSContiguousString(_unmanaged: self, range: range)
+    }
   }
+}
 
-  /*fileprivate*/ internal // TODO: private in Swift 4
-  init(_ s: SmallCocoaString) {
-    self.init(_bridgeObject(fromTagged: _smallStringMask), s.taggedPointer)
+extension _StringGuts {
+  @_versioned
+  internal
+  var _underlyingCocoaString: _CocoaString? {
+    if _fastPath(_isNative) {
+      return _nativeRawStorage
+    } else if _isNonTaggedCocoa {
+      return _nonTaggedCocoaObject
+    } else if _isSmallCocoa {
+      return _taggedCocoaObject
+    } else {
+      return nil
+    }
   }
+}
 
-  //
-  // Small Strings
-  //
-  ///*fileprivate*/ internal // TODO: private in Swift 4
-  public // TODO(StringGuts): for testing only
-  var _isError: Bool { return false }
+extension _StringGuts {
+  // Opaque strings
 
-  /*fileprivate*/ internal // TODO: private in Swift 4
-  var _error: FatalErrorString? {
-    guard _isError else { return nil }
-    return FatalErrorString()
+  @_inlineable
+  @_versioned
+  internal
+  var _isContiguous: Bool {
+    @inline(__always) get {
+      return !_isOpaque
+    }
   }
 
-  /*fileprivate*/ internal // TODO: private in Swift 4
-  init(_ s: FatalErrorString) {
-    fatalError("I AM ERROR")
+  @_inlineable
+  @_versioned
+  internal
+  var _isOpaque: Bool {
+    @inline(__always) get {
+      return _isSmallCocoa || _isNoncontiguousCocoa
```

Comment:
If we store the count for CocoaString, we'll probably end up just using a bit for isContiguous